MarsRover/Rover.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace MarsRover
+{
+    public class Rover
+    {
+        private Grid _grid;
+        private IHeading _heading;
+        private List<Obstacle> _obstacles;
+        private bool _obstacleDetected;
+
+        public Rover(Grid grid, IHeading heading, List<Obstacle> Obstacles)
+        {
+            _grid = grid.GridEntryAndSizeValidation(grid);
+            _heading = heading;
+            _obstacles = Obstacles;
+            _obstacleDetected = false;
+        }
+
+        public void TurnLeft() => _heading = _heading.TurnLeft();
+
+        public void TurnRight() => _heading = _heading.TurnRight();
+
+        public void MoveForward() => _grid = _heading.MoveForward(_grid);
+
+        public void MoveBackward() => _grid = _heading.MoveBackward(_grid);
+
+        public void InvalidCommand() { }
+
+        public string Direction() => _heading.GetHeading();
+
+        public int XCoordinate() => _grid.XCoordinate();
+
+        public int YCoodinate() => _grid.YCoordinate();
+
+        public int GridHeight() => _grid.GridHeight();
+
+        public int GridWidth() => _grid.GridWidth();
+
+        public bool ObstacleDetected() => _obstacleDetected;
+
+        public Rover ExecuteCommands(string commands)
+        {
+            foreach (var letter in commands)
+            {
+                var inputCommand = new InputCommands(letter.ToString());
+                var commandToRun = inputCommand.ValidateAndReturnCommand();
+                var commandType = this.GetType().Assembly.GetType(commandToRun);
+                var command = (ICommand)Activator.CreateInstance(commandType, this);
+                command.Execute();
+
+                if (_obstacleDetected = ObstaclesFound(XCoordinate(), YCoodinate()) == true) { return this; }
+            }
+            return this;
+        }
+
+        private bool ObstaclesFound(int xCoordinate, int yCoordinate)
```

Comment:
I would rename this to be singular since it only tells us if a single obstacle was found or not: `ObstacleFound`.  Also, we can greatly simplify this to be one line long by using LINQ's `any` function.